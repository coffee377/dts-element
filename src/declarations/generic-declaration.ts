import * as ts from 'typescript';
import {IType} from '../collections';
import {ElementKind} from '../constants';
import {create_element, IElement} from '../element';
import {transform} from '../transform';

export interface IGenericDeclarationOptions {
  name: string;
  extends?: IType;
  defalut?: IType;
}

export interface IGenericDeclaration
  extends IElement<ElementKind.GenericDeclaration>, IGenericDeclarationOptions {}

export const create_generic_declaration = (options: IGenericDeclarationOptions): IGenericDeclaration => ({
  ...create_element(ElementKind.GenericDeclaration),
  ...options,
});

export const transform_generic_declaration = (element: IGenericDeclaration, path: IElement<any>[]) =>
  ts.createTypeParameterDeclaration(
    element.name,
    element.extends && (transform(element.extends, [...path, element]) as ts.TypeNode),
    element.defalut && (transform(element.defalut, [...path, element]) as ts.TypeNode),
  );
