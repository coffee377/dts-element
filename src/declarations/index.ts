export * from './class-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './interface-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
