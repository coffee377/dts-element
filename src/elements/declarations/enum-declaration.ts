import {emit_declare} from '../../helpers/emit-declare';
import {emit_enum_members} from '../../helpers/emit-enum-members';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {EnumMember} from '../members/enum-member';

export interface IEnumDeclarationRequiredParameters {
  name: string;
}

export interface IEnumDeclarationOptionalParameters {
  members: EnumMember[];
}

export class EnumDeclaration
    extends Declaration<IEnumDeclarationRequiredParameters, IEnumDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IEnumDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      members: [],
    });
  }

  public _emit_raw(stack: Stack): string {
    const {name, members} = this.parameters;
    const declare = emit_declare(stack);
    return `${declare}enum ${name} ${emit_enum_members(members, stack)}`;
  }

}
