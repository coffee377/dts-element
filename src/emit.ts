import * as ts from 'typescript';
import {ElementKind} from './constants';
import {IElement} from './element';
import {emit_triple_slash_reference, ITripleSlashReference} from './others/triple-slash-reference';
import {transform} from './transform';

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) => {
  switch (element.kind) {
    case ElementKind.TripleSlashReference:
      return emit_triple_slash_reference(element as ITripleSlashReference);
    default:
      const ts_node = transform(element);
      const source_file = ts.createSourceFile('', '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
      const printer = ts.createPrinter(options);
      return printer.printNode(ts.EmitHint.Unspecified, ts_node, source_file);
  }
};
