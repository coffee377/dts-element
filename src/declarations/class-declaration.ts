import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {IClassMember} from '../members';
import {IIndexSignature} from '../others/index-signature';
import {transform} from '../transform';
import {IGeneralType} from '../types/general-type';
import {add_declare_modifier_if_need, create_type_nodes, create_type_parameters} from '../utils';
import {IGenericDeclaration} from './generic-declaration';

export interface IClassDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  members?: (IClassMember | IIndexSignature)[];
  abstract?: boolean;
  extends?: IGeneralType;
}

export interface IClassDeclaration
  extends IElement<ElementKind.ClassDeclaration>, IClassDeclarationOptions {}

export const create_class_declaration = (options: IClassDeclarationOptions): IClassDeclaration => ({
  ...create_element(ElementKind.ClassDeclaration),
  ...options,
});

// tslint:disable:ter-indent max-line-length

export const transform_class_declaration = (element: IClassDeclaration, path: IElement<any>[]) =>
  ts.createClassDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
                            [
                              ...(element.export === true)
                                ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                                : [],
                              ...(element.abstract === true)
                                ? [ts.createToken(ts.SyntaxKind.AbstractKeyword)]
                                : [],
                            ],
                            path,
                          ),
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* heritageClauses */ (element.extends === undefined)
                            ? []
                            : [ts.createHeritageClause(
                            /* token */ ts.SyntaxKind.ExtendsKeyword,
                            /* types */ [ts.createExpressionWithTypeArguments(
                                          /* typeArguments */ create_type_nodes(element.extends.generics, path),
                                          /* expression    */ ts.createIdentifier(element.extends.name),
                                        )],
                            )],
    /* members         */ (element.members || []).map(member => transform(member, path) as ts.ClassElement),
  );
