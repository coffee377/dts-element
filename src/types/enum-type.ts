import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {IEnumDeclaration} from '../declarations/enum-declaration';
import {create_element, IElement, IElementOptions} from '../element';

export interface IEnumTypeOptions extends IElementOptions {
  name: string | IEnumDeclaration;
}

export interface IEnumType
  extends IElement<ElementKind.EnumType>, IEnumTypeOptions {}

export const create_enum_type = (options: IEnumTypeOptions): IEnumType => ({
  ...create_element(ElementKind.EnumType),
  ...options,
});

export const transform_enum_type = (element: IEnumType, path: IElement<any>[]) =>
  ts.createTypeReferenceNode(
    /* typeName      */ (typeof element.name === 'string')
                          ? element.name
                          : element.name.name,
    /* typeArguments */ undefined,
  );
