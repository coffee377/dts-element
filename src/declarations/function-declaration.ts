import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement} from '../element';
import {transform} from '../transform';
import {create_function_type, IFunctionType} from '../types/function-type';

export interface IFunctionDeclarationOptions {
  name: string | undefined;
  type?: IFunctionType;
  export?: boolean;
}

export interface IFunctionDeclaration
  extends IElement<ElementKind.FunctionDeclaration>, IFunctionDeclarationOptions {}

export const create_function_declaration = (options: IFunctionDeclarationOptions): IFunctionDeclaration => ({
  ...create_element(ElementKind.FunctionDeclaration),
  ...options,
});

export const transform_function_declaration = (element: IFunctionDeclaration, path: IElement<any>[]) => {
  const function_type = transform(element.type || create_function_type(), path) as ts.FunctionDeclaration;
  return ts.createFunctionDeclaration(
    /* decorators     */ undefined,
    /* modifiers      */ (element.export === true)
                           ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                           : undefined,
    /* asteriskToken  */ undefined,
    /* name           */ element.name,
    /* typeParameters */ function_type.typeParameters,
    /* parameters     */ function_type.parameters,
    /* type           */ function_type.type,
    /* body           */ undefined,
  );
};
