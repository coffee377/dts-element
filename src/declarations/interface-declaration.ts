import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {transform} from '../transform';
import {IGeneralType} from '../types/general-type';
import {create_object_type, IObjectType} from '../types/object-type';
import {add_declare_modifier_if_need, create_type_arguments, create_type_parameters} from '../utils';
import {IGenericDeclaration} from './generic-declaration';

export interface IInterfaceDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  type?: IObjectType;
  extends?: IGeneralType[];
}

export interface IInterfaceDeclaration
  extends IElement<ElementKind.InterfaceDeclaration>, IInterfaceDeclarationOptions {}

export const create_interface_declaration = (options: IInterfaceDeclarationOptions): IInterfaceDeclaration => ({
  ...create_element(ElementKind.InterfaceDeclaration),
  ...options,
});

// tslint:disable:ter-indent max-line-length

export const transform_interface_declaration = (element: IInterfaceDeclaration, path: IElement<any>[]) =>
  ts.createInterfaceDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
                            (element.export === true)
                              ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                              : undefined,
                            path,
                          ),
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* heritageClauses */ element.extends && [ts.createHeritageClause(
                            /* token */ ts.SyntaxKind.ExtendsKeyword,
                            /* types */ element.extends.map(interface_type => ts.createExpressionWithTypeArguments(
                                          /* typeArguments */ create_type_arguments(interface_type.generics, path),
                                          /* expression    */ ts.createIdentifier(interface_type.name),
                                        )),
                          )],
    /* members         */ (transform(element.type || create_object_type(), path) as ts.TypeLiteralNode).members,
  );
