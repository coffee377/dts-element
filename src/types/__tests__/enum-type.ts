import {create_enum_declaration} from '../../declarations/enum-declaration';
import {emit} from '../../emit';
import {create_enum_type} from '../enum-type';

it('should return correctly with name (string)', () => {
  expect(emit(
    create_enum_type({
      name: 'E',
    }),
  )).toMatchSnapshot();
});

it('should return correctly with name (EnumDeclaration)', () => {
  expect(emit(
    create_enum_type({
      name: create_enum_declaration({name: 'E'}),
    }),
  )).toMatchSnapshot();
});
