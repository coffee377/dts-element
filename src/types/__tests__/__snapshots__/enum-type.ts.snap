// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with name (EnumDeclaration) 1`] = `"E"`;

exports[`should return correctly with name (string) 1`] = `"E"`;
